src/middlewared/middlewared/plugins/acm.py

```diff
@@ -0,0 +1,736 @@
+                            'acme_authorization.domain',
+                            f'Error answering challenge for {domain} : {e}'
+                        )
+
+            finally:
+                job.set_progress(
+                    progress,
+                    f'DNS challenge {"completed" if status else "failed"} for {domain}'
+                )
+
+        if verrors:
+            raise verrors
+
+    @periodic(86400, run_on_start=True)
+    @job(lock='acme_cert_renewal')
+    def renew_certs(self, job):
+        certs = self.middleware.call_sync('acme.query')
+        progress = 0
+        for cert in certs:
+            # TODO: Make the renew date customizable, perhaps add a field to the cert model
+            # TODO: Make renewal optional with default set as True
+            progress += (100 / len(certs))
+
+            if (pytz.utc.localize(cert['expire']) - datetime.datetime.now(datetime.timezone.utc)).days < 10:
+                # renew cert
+                self.logger.debug(f'Renewing certificate {cert["name"]}')
+                final_order = self.issue_certificate(
+                    job, progress / 4, {
+                        'tos': True,
+                        'directory_uri': cert['acme']['directory'],
+                        'domain_dns_mapping': cert['domain_authenticators']
+                    },
+                    cert
+                )
+
+                self.middleware.call_sync(
+                    'datastore.update',
+                    self._config.datastore,
+                    cert['id'],
+                    {
+                        'certificate': final_order.fullchain_pem,
+                        'acme_uri': final_order.uri,
+                        'expire': final_order.body.expires.astimezone(pytz.utc).strftime("%Y-%m-%d"),
+                        'chain': True if len(RE_CERTIFICATE.findall(final_order.fullchain_pem)) > 1 else False,
+                    },
+                    {'prefix': self._config.datastore_prefix}
+                )
+
+            job.set_progress(progress)
+
+    @job(lock='acme_cert_revoke')
+    def do_delete(self, job, id):
+        # First we revoke the certificate and then delete it from the system
+        cert = self.middleware.call_sync(
+            'acme.query',
+            [['id', '=', id]]
+        )
+
+        verrors = ValidationErrors()
+
+        if not cert:
+            verrors.add(
+                'acme_revoke.id',
+                'Kindly provide a valid ACME based certificate id'
+            )
+        else:
+            cert = cert[0]
+
+        if verrors:
+            raise verrors
+
+        client, key = get_acme_client_and_key(self.middleware, cert['acme']['directory'], True)
+
+        job.set_progress(60)
+
+        try:
+            client.revoke(
+                jose.ComparableX509(crypto.load_certificate(crypto.FILETYPE_PEM, cert['certificate'])),
+                0
+            )
+        except errors.ClientError as e:
```

Comment:
If revoke fails user will never be able to delete certificate from UI?

Reply:
For now yes. What do you think is the best way to move ahead with this ? Should we delete a cert which is still recognized by ACME server ?

Reply:
I believe so, perhaps provide a force option that go ahead and delete from database even if revoke fails for some reason?